# FUEL DATA ANALYSIS REPORT

In [1]:
import pandas as pd
import numpy as np

#### Question 1: If you're given two lists: A = [1,2,3,4,5,6], B = [13,21,34]. The tasks is to create a list with the elements of A and B in a single dimension with output: A_B = [1,2,3,4,5,6,13,21,34]. Which of the following option is the best way too create this list?


In [85]:
A = [1,2,3,4,5,6]
B = [13,21,34]
print(A)
print(B)

[1, 2, 3, 4, 5, 6]
[13, 21, 34]


In [86]:
A_B = A.extend(B)
print(A)

[1, 2, 3, 4, 5, 6, 13, 21, 34]


A.append(B) creates an updated list out of the two list of arrays.

In [68]:
A_B = A + B
print(A_B)

[1, 2, 3, 4, 5, 6, 13, 21, 34]


#### Question 2: How do you create an identity matrix in python?

In [4]:
C  = np.identity(3)
print(C)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


Among the underlisted procedures, np.identity(3) best defines an identity matrix.

In [5]:
fuel_data = pd.read_csv(r'C:\Users\Gbenga\Desktop\fuel_ferc1.csv')

In [6]:
fuel_data.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [7]:
fuel_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     29523 non-null  object 
 1   utility_id_ferc1              29523 non-null  int64  
 2   report_year                   29523 non-null  int64  
 3   plant_name_ferc1              29523 non-null  object 
 4   fuel_type_code_pudl           29523 non-null  object 
 5   fuel_unit                     29343 non-null  object 
 6   fuel_qty_burned               29523 non-null  float64
 7   fuel_mmbtu_per_unit           29523 non-null  float64
 8   fuel_cost_per_unit_burned     29523 non-null  float64
 9   fuel_cost_per_unit_delivered  29523 non-null  float64
 10  fuel_cost_per_mmbtu           29523 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


#### Question 3: Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [69]:
fuel_data.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean()

fuel_type_code_pudl
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: fuel_cost_per_unit_burned, dtype: float64

The fuel type code with the lowest average fuel cost per unit burned is Gas with the value of 13.659397.

#### Question 4: What is the standard deviation and 75th percentile of the measure of energy per unit in two decimal places?

In [71]:
fuel_data['fuel_mmbtu_per_unit'].describe()

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

The measure of fuel energy per unit standard deviation is 10.60 while 75th percentile is 17.01

#### Question 5: What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [73]:
print(fuel_data['fuel_qty_burned'].skew())
print(fuel_data['fuel_qty_burned'].kurt())

15.851495469109503
651.3694501337732


Answer: Skewness: 15.85, Kurtosis: 651.37

#### Question 6: What feature has missing values and what is the total number of missing value and percentage of the missing rows as a factor of the total number of rows in three decimal places?

In [74]:
fuel_data.isna().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

Answer: Feature: fuel_unit, 
        Total: 180, 
        Percent: 0.610

#### Question 7: The feature with missing values falls under what category? What missing value imputation technique would you use?

In [75]:
fuel_data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

Feature of missing value is categorical and mode imputation. Mode imputation is used as a result of it being the highest type in the population.

#### Question 8: Which of the features has the second and third lowest correlation with the fuel cost per unit burned?

In [76]:
fuel_data.corr()

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

Fuel cost per unit burned and fuel quantity burned (-0.019) - second lowest value;
fuel cost per unit burned and fuel energy per unit (-0.01) - third lowest value

#### Question 9: For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [77]:
fuel_data.groupby(['fuel_type_code_pudl','report_year'])['fuel_type_code_pudl','fuel_cost_per_unit_burned'].sum()

C:\Users\Dnear\anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


fuel_cost_per_unit_burned
fuel_type_code_pudl report_year                           
coal                1994                         14984.572
                    1995                         14571.785
                    1996                         12694.803
                    1997                         43742.178
                    1998                         11902.597
...                                                    ...
waste               2014                           591.594
                    2015                           629.270
                    2016                           539.079
                    2017                           452.079
                    2018                           660.972

[148 rows x 1 columns]

The percentage change in the fuel cost per unit burned in 1998 compared to 1994 is -21%. That is (11902.597 - 14984.572)/14984.572 * 100

In [78]:
 (11902.597 - 14984.572)/14984.572 * 100

-20.567654518260518

#### Question 10: Which year has the highest average fuel cost per unit delivered?

In [80]:
fuel_data.groupby('report_year')['fuel_cost_per_unit_delivered'].mean()

report_year
1994       63.636060
1995       32.735269
1996     9196.705948
1997    11140.197239
1998      287.154420
1999       25.551627
2000      985.362877
2001       60.050396
2002       47.594361
2003       55.663493
2004      139.524275
2005       41.438184
2006       38.657484
2007       43.325023
2008       58.588197
2009      652.694163
2010       91.862105
2011       59.774667
2012       60.994502
2013      172.307591
2014      192.737183
2015      326.535511
2016      103.901761
2017       46.196861
2018      499.269966
Name: fuel_cost_per_unit_delivered, dtype: float64

The highest average fuel cost per unit delivered is in year 1997 and its average value is 11140.197.